client/securedrop_client/gui/widgets.py

```diff
@@ -424,6 +429,139 @@ def clear_message(self) -> None:
+    "Delete Sources" (batch-delete) action.
+
+    For user-facing naming consistency, these items won't be called
+    "batch/bulk <delete>", but simply "<verb> <noun>s" (eg "Delete Sources"), where
+    the original nomenclature comes from the individual Source overflow QAction menu
+    items. Each item may have a tooltip, visible on hover, that provides a more
+    lengthy explanation (e.g., "Delete multiple source accounts").
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.setObjectName("BatchActionToolbar")
+        self.setContentsMargins(0, 0, 0, 0)
+
+        palette = QPalette()
+        palette.setBrush(
+            QPalette.Background, QBrush(Qt.NoBrush)
+        )  # This makes the widget transparent
+        self.setPalette(palette)
+
+        # Style and attributes
+        self.setMovable(False)
+        self.setToolButtonStyle(Qt.ToolButtonStyle.ToolButtonTextBesideIcon)
+        self.setSizePolicy(QSizePolicy.Preferred, QSizePolicy.Fixed)
+
+        self.delete_sources_action = QAction(
+            QIcon(load_image("delete_sources_toolbar_icon.svg")), _("DELETE SOURCES"), self
+        )
+        self.delete_sources_action.setObjectName("BatchActionButton")
+        self.delete_sources_action.setToolTip(
+            _(
+                "Delete selected source accounts. "
+                "Ctrl+click sources below to select multiple sources."
+            )
+        )
+        self.delete_sources_action.triggered.connect(self.on_action_triggered)
+        self.button = self.widgetForAction(self.delete_sources_action)
+
+        # Add spacer. "select all" checkbox can replace the spacer in future
+        spacer = QWidget()
+        spacer.setFixedSize(14, 14)  # sourcewidget spacer size
+        self.addWidget(spacer)
+        self.addAction(self.delete_sources_action)
+
+    def setup(self, controller: Controller) -> None:
+        self.controller = controller
+
+    def hide_action(self) -> None:
+        self.delete_sources_action.setVisible(False)
+
+    def show_action(self) -> None:
+        self.delete_sources_action.setVisible(True)
+
+    @pyqtSlot()
+    def on_action_triggered(self) -> None:
+        if self.controller.api is None:
+            self.controller.on_action_requiring_login()
+        else:
+            # The current source selection is continuously received by the controller
+            # as the user selects and deselects; here we retrieve the selection
+            targets = self.controller.get_selected_sources()
+            if targets:
+                dialog = DeleteSourceDialog(targets)
+                dialog.accepted.connect(lambda: self.controller.delete_sources(targets))
+                dialog.exec()
+            else:
+                # No selected sources should return an empty set, not None
+                logger.error("Toolbar action triggered without valid data from controller.")
```

Comment:
Ah, thanks, too hasty of a refactor, I'll push that change.

Yes, we should display a dialog for the empty list for now -  initial discussion [here](https://github.com/freedomofpress/securedrop-client/issues/2160#issue-2461049651).   I think the options are: 

- Display a flashed toast/growl message (best case, but slightly more implementation work, maybe future improvement) telling user "You must select some sources".   This kind of improvement could also be added at a time when we add other improvements/enhancements (such as: checkmark if we get feedback we need it, counter of the number of selected sources, select all/none capability, or another batch action) 
- Display the dialog but with a generic "Nothing selected" popup (current behaviour). It's not beautiful but it provides clear user feedback
- Do nothing - don't like this option, the user won't know if it's because they mis-clicked, the button is broken, or why. 

I opted for the least implementation work + most user feedback option in the interests of time, but I agree it's a bit kludgy.